src/Workspaces/Core/MSBuild/MSBuild/BuildHostProcessManager.cs

```diff
@@ -135,7 +135,17 @@ private void BuildHostProcess_Disconnected(object? sender, EventArgs e)
 
     public async ValueTask DisposeAsync()
     {
-        foreach (var process in _processes.Values)
+        List<BuildHostProcess> processesToDispose;
+
+        // Copy the list out while we're in the lock, otherwise as we dispose these events will get fired, which
+        // may try to mutate the list while we're enumerating.
+        using (await _gate.DisposableWaitAsync().ConfigureAwait(false))
+        {
+            processesToDispose = _processes.Values.ToList();
```

Comment:
is the addition to _processes also under this lock?

Reply:
Yes, see BuildHostProcessManager.GetBuildHostAsync.